### ISSUE TITLE
Add GPU-GPU interfaces

### ISSUE THREAD

**Issue by AlexVaq** [lattice/quda#278]
Aof now, QUDA is mostly oriented towards a CPU-GPU interface, where the data is read from CPU, used to compute something in the GPU and then written back in CPU. There are, though, some applications that require data generated in the GPU to stay in the GPU for further calculations. An example are the bilinear contractions and inversion of sources, which should be coupled together to avoid intermediate and useless CPU transfers. That's why it is necessary to develop some kind of GPU-GPU interfaces that hardly communicate to host.

**Comment by mikeaclark** [comment]
Initial support for post-solve resident quark fields has been added to `invertMultishiftQuda` and `cloverForceQuda`.  There are new fields in `QudaInvertParam` that can instrument this:
```
  // store the solution(s) in std::vector<cudaColorSpinorField*> solutionResident
  int make_resident_solution; 
  // use the solution(s) in std::vector<cudaColorSpinorField*> solutionResident 
  int use_resident_solution; 
```
I used a std::vector of solution vectors so that it can easily handle the solutions out of the multi-shift solver.  It will be trivial to add support for this to the regular `invertQuda` solver.

@AlexVaq does this approach meet your needs?
